### PR TITLE
Add "AvailabilityZone:all" parameter to IP targets when TargetGroup is outside of Controller's VPC

### PR DESCRIPTION
**Pull request by a2ush**
### Issue

#2650

### Description

Currently, when you deploy targetgroupbinding with ALB/NLB TargetGroup which is outside of Loadbalancer Controller's VPC, the Controller cannot do RegisterTargets API because it doesn't add "AvailabilityZone:all" parameter to IP targets.

[Loadbalancer Controller Log]
```
Ex)
{"level":"error","ts":1658843126.3276293,"logger":"controller-runtime.manager.controller.targetGroupBinding","msg":"Reconciler error","reconciler group":"elbv2.k8s.aws","reconciler kind":"TargetGroupBinding","name":"my-tgb","namespace":"default","error":"ValidationError: The Availability Zone is required for IP address '192.168.28.187' because it is not in the VPC\n\tstatus code: 400, request id: b228b6b8-..."}
```
https://docs.aws.amazon.com/elasticloadbalancing/latest/APIReference/API_TargetDescription.html
 > If the target type is ip and the IP address is in a subnet of the VPC for the target group, the Availability Zone is automatically detected and this parameter is optional. If the IP address is outside the VPC, this parameter is required.

This PR helps Loadbalancer Controller to add "AvailabilityZone:all" parameter to IP targets in this situation, by comparing TargetGroup's VPC with Loadbalancer Controller's VPC.
That is, you will be able to register pod's IP address to other VPC's TargetGroups.

New Loadbalancer Controller's log
```
Ex)
{"level":"info","ts":1658812216.264621,"msg":"registering targets","arn":"arn:aws:elasticloadbalancing:<Region>:<AccountID>:targetgroup/my-tgb/1234567890","targets":[{"AvailabilityZone":"all","Id":"192.168.36.77","Port":80},{"AvailabilityZone":"all","Id":"192.168.43.133","Port":80},{"AvailabilityZone":"all","Id":"192.168.53.216","Port":80}]}
{"level":"info","ts":1658812216.5066864,"msg":"registered targets","arn":"arn:aws:elasticloadbalancing:<Region>:<AccountID>:targetgroup/my-tgb/1234567890"}
```

### Checklist
- [ ] Added tests that cover your change (if possible)
- [ ] Added/modified documentation as required (such as the `README.md`, or the `docs` directory)
- [X] Manually tested
- [X] Made sure the title of the PR is a good description that can go into the release notes

### BONUS POINTS checklist: complete for good vibes and maybe prizes?! :exploding_head:
- [ ] Backfilled missing tests for code in same general area :tada:
- [ ] Refactored something and made the world a better place :star2:

### CODE REVIEW

**Comment by daghsentinel** [comment]
Any chance of getting anyone to look into this? I am having the exact same problems as staed in the issue

**Comment by tmehlinger** [comment]
I'm also running into this. I'd really appreciate someone reviewing and merging it.

File: pkg/targetgroupbinding/resource_manager.go [hunk 1]
```diff
@@ -382,6 +382,10 @@ func (m *defaultResourceManager) registerPodEndpoints(ctx context.Context, tgARN
 	if err != nil {
 		return err
 	}
+	tgInfo, err := m.targetsManager.DescribeTargetGroup(ctx, tgARN)
```
Comment (kishorj) [inline pkg/targetgroupbinding/resource_manager.go:385]: I'd prefer caching the target group vpc Id so we don't incur the DescribeTargetGroup call with every register.

Comment (M00nF1sh) [reply pkg/targetgroupbinding/resource_manager.go:385]: +1, we should minimize DescribeTargetGroup calls in this critical path.

we can either add cache or mutate the targetGroupBinding spec during MutationWebhook to capture vpcID in spec/annotation.

I prefer to have a cache for now given its simper and will be a two way door.

**Comment by tmehlinger** [comment]
@kishorj thanks for checking this out, much appreciated!

File: pkg/targetgroupbinding/resource_manager.go (hunk 1 above)
Comment (a2ush) [reply pkg/targetgroupbinding/resource_manager.go:385]: Thank you for your advice!
I agree that. But I'm afraid I'm not sure how to best implement a cache for the controller.

Could you give me some reference information (e.g. similar implement, sample code)?

**Comment by pescetto** [comment]
Thanks for your work on this, any idea when this bug fix might get merged in? We're running into this also and can't use aws-load-balancer-controller until this gets resolved.

**Pull request closed without merging**